core/local/layers/linux.js

```diff
@@ -0,0 +1,180 @@
+        return
+      }
+      this.next.process(batch)
+      for (const event of batch) {
+        if (event.doc.docType === 'folder') {
+          await this.watch(event.doc.path)
+        }
+      }
+    } catch (err) {
+      // The directory may been removed since we wanted to watch it
+    }
+  }
+
+  async process (events /*: AtomWatcherEvent[] */) {
+    // TODO logger
+    // TODO ignore
+    const batch /*: LayerEvent[] */ = []
+    for (const event of events) {
+      switch (event.action) {
+        case 'created':
+          const eAdd = await this.buildAddEvent(event.path)
+          batch.push(eAdd)
+          if (eAdd.doc.docType === 'folder') {
+            this.watch(eAdd.doc.path)
+          }
+          break
+        case 'modified':
+          batch.push(await this.buildUpdateEvent(event.path))
+          break
+        case 'deleted':
+          const eDel = await this.buildRemoveEvent(event.path, event.kind)
+          batch.push(eDel)
+          const w = this.watchers.get(eDel.doc.path)
+          if (w) {
+            w.dispose()
+            this.watchers.delete(eDel.doc.path)
+          }
+          break
+        case 'renamed':
+          batch.push(await this.buildMoveEvent(event.path, event.oldPath))
+          break
+        default:
+          throw new Error(`Unknown atom/watcher action ${event.action}`)
+      }
+    }
+    this.next.process(batch)
+  }
+
+  async buildAddEvent (abspath /*: string */) /*: Promise<LayerAddEvent> */ {
+    let doc /*: ?Metadata */
+    const fpath = this.relativePath(abspath)
+    const stats = await fse.stat(abspath)
+    if (stats != null && stats.isDirectory()) {
+      doc = buildDir(fpath, stats)
+    } else {
+      doc = buildFile(fpath, stats, '')
+    }
+    return { action: 'add', abspath, doc }
+  }
+
+  async buildUpdateEvent (abspath /*: string */) /*: Promise<LayerUpdateEvent> */ {
+    let doc /*: ?Metadata */
+    const fpath = this.relativePath(abspath)
+    const stats = await fse.stat(abspath)
+    if (stats != null && stats.isDirectory()) {
+      doc = buildDir(fpath, stats)
+    } else {
+      doc = buildFile(fpath, stats, '')
+    }
+    return { action: 'update', abspath, doc }
+  }
+
+  async buildRemoveEvent (abspath /*: string */, kind /*: string */) /*: Promise<LayerRemoveEvent> */ {
+    const fpath = this.relativePath(abspath)
+    let doc /*: ?Metadata */
+    if (kind === 'directory') {
+      doc = buildDir(fpath, new fs.Stats())
+    } else {
+      doc = buildFile(fpath, new fs.Stats(), '')
+    }
+    return { action: 'remove', abspath, doc }
+  }
+
+  async buildMoveEvent (abspath /*: string */, oldpath /*: string */) /*: Promise<LayerMoveEvent> */ {
+    let doc /*: ?Metadata */
+    let src /*: ?Metadata */
+    const fpath = this.relativePath(abspath)
+    const stats = await fse.stat(abspath)
+    if (stats != null && stats.isDirectory()) {
+      doc = buildDir(fpath, stats)
+      src = buildDir(oldpath, new fs.Stats())
```

Comment:
I'm not comfortable with the use of empty stats, the dir (or file) ends up with an invalid date, flow doesn't even complain, neither does the `Date` constructor, everything would just crash at some point later when calling whatever Date method...

I assume we end up fixing everything in another layer, but it still feels weird to me...

Also the more I think about it, the more I believe we're currently trying to build docs too early even in the current code, which leads to a lot of error-prone *fixing* afterwards...
Shouldn't we just build a data structure that represents the event identified by the watcher, then derive a doc from it later in merge?

Reply:
Also the more I think about it, the more I believe we're currently trying to build docs too early even in the current code, which leads to a lot of error-prone *fixing* afterwards...
Shouldn't we just build a data structure that represents the event identified by the watcher, then derive a doc from it later in merge?

Reply:
I feel the pain, but I don't have a good solution for now.

> Shouldn't we just build a data structure that represents the event identified by the watcher, then derive a doc from it later in merge?

I don't see what it will change. This data structure that represents the event will still be build step by step, and we will still juggle to be sure it has the expected properties. It's just moving the issue to another structure, no?

Flow seems too weak for our step by step building of events (or maybe I don't understand it enough).

Reply:
We could pass the original atom-watcher event to the various `#build*Event()` methods without having to destructure it to later restructure it into something almost similar.
From what I can see, the next layer basically needs the atom-watcher event + stats.
Everything else can actually be derived from those two.
So I find the current code a bit complicated regarding the purpose...
Or did I miss something?

Reply:
TLDR: pass an AtomWatcherEvent to every `#build*Event()` method and pass a batch of `{atomWatcherEvent, relpath, stats}` (without any doc) to the next layer?

Reply:
(I'm not sure a `#build*Event()` is actually needed for every action type)

Reply:
I could use a custom struct for the events in the layer, and transform them to something compatible with Merge in the last layer (Dispatcher), but this struct will still be build step by step. It just moves complexity, not fix it.

Reply:
Could checksums be computed in a single async function doing a Promise.reduce of the file add/update event paths over a Map<Path,Md5sum> using the checksumer?
Could stats be computed the same way?
With those 2 maps and the original event, couldn't we build a full doc at once?